### PR TITLE
SLING-9593: Abstract out binary handling to allow plugging of differe…

### PR DESCRIPTION
**Pull request by amit-jain**
…nt stores

- Introduced an interface BinaryStore and JcrBinaryStore as the default implementation storing the binaries in JCR with no change in the existing strategy of storing based on size.

### CODE REVIEW

**Comment by amit-jain** [comment]
Depends on the interface https://github.com/apache/sling-org-apache-sling-distribution-journal-messages/pull/4

**Comment by amit-jain** [comment]
Also depends on PR https://github.com/apache/sling-org-apache-sling-distribution-journal/pull/55.

Will update the PR once the dependent PRs have been merged

**Comment by cschneider** [comment]
I merged the new interface in messages and also merged my PR for the PackageRepo changes. Can you adapt your PR to these?

File: src/main/java/org/apache/sling/distribution/journal/impl/publisher/PackageMessageFactory.java [hunk 1]
```diff
@@ -116,27 +112,18 @@ private PackageMessage createAdd(DistributionPackageBuilder packageBuilder, Reso
...
+            LOG.info("Creating package binary with id [{}] for package [{}], length [{}]", id, disPkg.getId(), pkgLength);
+            storeRef =  binaryStore.put(id, disPkg.createInputStream(), pkgLength);
+        } catch (IOException e) {
+            throw new DistributionException(e.getMessage(), e);
+        }
+
+        if (StringUtils.isNotEmpty(storeRef)) {
+            pkgBuilder.pkgBinaryRef(storeRef);
         } else {
             pkgBuilder.pkgBinary(pkgBinary);
```
Comment (cschneider) [inline src/main/java/org/apache/sling/distribution/journal/impl/publisher/PackageMessageFactory.java:128]: On top of the method we always create the pkgBinary. I propose we only create it when the binaryStore does not return a storeRef.

Comment (amit-jain) [reply src/main/java/org/apache/sling/distribution/journal/impl/publisher/PackageMessageFactory.java:128]: @cschneider Yes, its a wasted effort to put the binry in memory but the important thing is that the length is being calculated from that and being used to build the PackageMessageBuilder. I couldn't find out for sure whether the DistributionPackage.getSize() always returns the length/size.

Comment (cschneider) [reply src/main/java/org/apache/sling/distribution/journal/impl/publisher/PackageMessageFactory.java:128]: Ah .. sorry .. did not see this .. in this case it is fine.

**Pull request merged**